### PR TITLE
Add support for msgctxt to i18n

### PR DESCRIPTION
**Pull request by drd**
This is a bit of a blast from the past, but some internet meandering led me to realize that the work I had done at Idealist many years ago to add support for a new `i18n:ctxt` attribute (think `pgettext`) never landed. The tests still appear to pass :)

Trac ticket: https://genshi.edgewall.org/ticket/580

### CODE REVIEW

**Comment by hodgestar** [comment]
@drd I know it's been awhile, but would you mind merging master into this branch and then I'll review.

Apologies for this issue taking seven year! :see_no_evil:

**Comment by spaceone** [comment]
@hodgestar FYI: in github PR's you can push to the branch if `allow edits from maintaines` is enabled for this PR. You can simply rebase the branch on `master`, there are only 3 simple conflicts.

**Comment by hodgestar** [comment]
@spaceone Thanks! I was actually hoping @drd could take a quick look again so that the code will have had two recent sets of eyes on it when it gets merged. If you're happy to review it with me, that would be awesome (but if not, no worries).

**Review by hodgestar — COMMENTED** [review]
Left a couple of small suggestions and I need to see if the tests are passing after my merge of the recent changes, but thank you for dusting this off and submitting it. Much appreciated.

File: genshi/filters/tests/i18n.py [hunk 1]
```diff
@@ -85,10 +85,53 @@ def ungettext(self, msgid1, msgid2, n):
...
+        try:
+            return self._catalog[(context, message)]
+        except KeyError:
+            if self._fallback:
+                return self._fallback.upgettext(context, message)
+            return unicode(message)
+
+    if not IS_PYTHON2:
+        pgettext = upgettext
+        del upgettext
```
Comment (hodgestar) [inline genshi/filters/tests/i18n.py:105]:
```suggestion
    if IS_PYTHON2:
        def upgettext(self, context, message):
            try:
                return self._catalog[(context, message)]
            except KeyError:
                if self._fallback:
                    return self._fallback.upgettext(context, message)
                return unicode(message)
    else:
        def pgettext(self, context, message):
            try:
                return self._catalog[(context, message)]
            except KeyError:
                if self._fallback:
                    return self._fallback.upgettext(context, message)
                return message
```

**Comment by hodgestar** [comment]
@drd Thank you for taking the time to resurrect this. I've cleaned it up a bit and am merging.

**Pull request merged**